explorer/puller/tests/facade/test_NemPullerFacade.py

```diff
@@ -0,0 +1,192 @@
+		)
+
+	def tearDown(self):
+		# Destroy the temporary PostgreSQL database
+		self.postgresql.stop()
+		self.temp_config_file.close()
+
+	def _create_facade_and_setup(self, return_values=None):
+		# Arrange:
+		# Creating facade and setting up the mock data
+		node_url = 'http://localhost:7890'
+		config_file = self.temp_config_file.name
+		facade = NemPullerFacade(node_url, config_file)
+
+		# Mock nem api endpoint data
+		if return_values:
+			for method, value in return_values.items():
+				setattr(facade._client, method, value)
+
+		return facade
+
+	@staticmethod
+	def _convert_memoryview_to_bytes(data):
+		# Convert memoryview objects to bytes
+		return [tuple(bytes(item) if isinstance(item, memoryview) else item for item in row) for row in data]
+
+	def test_initialization(self):
+		# Act:
+		facade = self._create_facade_and_setup()
+
+		# Assert:
+		self.assertIsInstance(facade.database(), NemDatabase)
+		self.assertIsInstance(facade.client(), NemClient)
+		self.assertEqual(facade.client().endpoint, 'http://localhost:7890')
+		self.assertIsNone(facade.nem_facade)
+
+	async def test_can_setup_facade(self):
+		# Arrange:
+		facade = self._create_facade_and_setup({
+			'node_network': AsyncMock(return_value='mainnet')
+		})
+
+		# Act:
+		await facade.setup_facade()
+
+		# Assert:
+		self.assertIsNotNone(facade.nem_facade)
+		self.assertEqual(str(facade.nem_facade.network), 'mainnet')
+
+	async def test_can_insert_nemesis_block(self):
+		# Arrange:
+		facade = self._create_facade_and_setup({
+			'node_network': AsyncMock(return_value='mainnet'),
+			'get_block': AsyncMock(return_value=CHAIN_BLOCK_1)
+		})
+
+		await facade.setup_facade()
+
+		with facade.database() as databases:
+			databases.create_tables()
+
+			# Act:
+			await facade.sync_nemesis_block()
+
+			# Assert:
+			cursor = facade.nem_db.connection.cursor()
+
+			cursor.execute('SELECT * FROM blocks WHERE height = %s', (2, ))
+			result = cursor.fetchone()
+
+			result = self._convert_memoryview_to_bytes([result])[0]
+
+			facade._client.get_block.assert_called_once_with(1)
+			self.assertEqual(result, self.expected_result_0)
+			self.assertIsNotNone(result)
+
+	async def test_can_insert_blocks(self):
+		# Arrange:
+		facade = self._create_facade_and_setup({
+			'node_network': AsyncMock(return_value='mainnet'),
+			'get_blocks_after': AsyncMock(return_value={
+				'data': [CHAIN_BLOCK_1, CHAIN_BLOCK_2]
+			})
+		})
+
+		await facade.setup_facade()
+
+		with facade.database() as databases:
+			databases.create_tables()
+
+			# Act:
+			await facade.sync_blocks(1, 2)
+
+			# Assert:
+			cursor = facade.nem_db.connection.cursor()
+
+			cursor.execute('SELECT * FROM blocks')
+			results = cursor.fetchall()
+
+			results = self._convert_memoryview_to_bytes(results)
+
+			facade._client.get_blocks_after.assert_called_once_with(1)
+			self.assertEqual(len(results), 2)
+			self.assertEqual(results[0], self.expected_result_0)
+			self.assertEqual(results[1], self.expected_result_1)
+
```

Comment:
you should have additional test, where sync_blocks fails.
you can simulate this by first calling sync_blocks(1,2) inside Arrange, and then calling it once more inside Act (i.e. with (2, 5).

the second call should fail (cause you already have height == 2 in db)

the reason for this test is to show that on failure no new blocks are added to db